projects/ADuCM3029_demo_cn0435/src/cli.c

```diff
@@ -0,0 +1,403 @@
+*    on or directly connected to an Analog Devices Inc. component.
+*
+* THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR IMPLIED
+* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT, MERCHANTABILITY
+* AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+* IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <string.h>
+#include "cli.h"
+#include "config.h"
+#if defined(CLI_INTEFACE)
+/******************************************************************************/
+/************************** Variable Definitions ******************************/
+/******************************************************************************/
+
+extern uint8_t uart_current_line[100];
+extern uint8_t uart_previous_line[100];
+extern uint8_t uart_line_index;
+extern uint8_t uart_cmd;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * Setup the CLI module of the application.
+ *
+ * Allocate memory for the driver handler and initialize the physical layer.
+ *
+ * @param [out] device    - Pointer to the device handler structure.
+ * @param [in] init_param - Pointer to the device initialization structure.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t cli_setup(struct cli_desc **device, struct cli_init_param *init_param)
+{
+	int32_t ret;
+	struct cli_desc *dev;
+
+	dev = calloc(1, sizeof *dev);
+	if (!dev)
+		return -1;
+
+	ret = usr_uart_init(&dev->uart_device, init_param->uart_init);
+	if(ret < 0)
+		goto error;
+
+	*device = dev;
+
+	return ret;
+error:
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * Free resources allocated by cli_setup().
+ *
+ * @param [in] dev - Pointer to the device descriptor.
+ *
+ * @return 0 if success, negative error code otherwise.
+ */
+int32_t cli_remove(struct cli_desc *dev)
+{
+	int32_t ret;
+
+	if(!dev)
+		return -1;
+
+	ret = usr_uart_remove(dev->uart_device);
+	if(ret < 0)
+		return ret;
+
+	free(dev);
```

Comment:
should we still free dev if uart_remove fails ?

Reply:
No